File: askdog-apps/iOS/AskDog/AskDog/Classes/Third/TopScrollMenuView/HYTopScrollMenuView.m
//
//  HYTopScrollMenuView.m
//  textMenu
//
//  Created by Hooying_zhaody on 16/9/19.
//  Copyright © 2016年 Hooying_zhaody. All rights reserved.
//

#import "HYTopScrollMenuView.h"

#define menuView_H          44
#define button_W            49

@interface HYTopScrollMenuView()<SGTopScrollMenuDelegate, UIScrollViewDelegate>
@property (nonatomic, strong) NSMutableArray *viewData;
@property (nonatomic, strong) NSArray *titles;
@end

@implementation HYTopScrollMenuView
{
    CGFloat _view_H;
    CGFloat _view_W;
}

- (NSMutableArray *)viewData {
    if (_viewData == nil) {
        _viewData = [[NSMutableArray alloc] init];
    }
    return _viewData;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        _view_W = CGRectGetWidth(frame);
        _view_H = CGRectGetHeight(frame);
    }
    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if(self){
        [self setupViews];
    }
    return self;
}

//- (id)awakeAfterUsingCoder:(NSCoder *)aDecoder {
//    self = [super awakeAfterUsingCoder:aDecoder];
//    if (self) {
//        _view_W = CGRectGetWidth(self.frame);
//        _view_H = CGRectGetHeight(self.frame);
//    }
//    return self;
//}

//- (void)awakeFromNib {
//    [super awakeFromNib];
//    _view_W = CGRectGetWidth(self.frame);
//    _view_H = CGRectGetHeight(self.frame);
//}

- (void)reloadTopScrollMenuView {
    
    [self.menuView removeFromSuperview];
    [self.bgScrollView removeFromSuperview];
    [self.viewData removeAllObjects];
    [self setupViews];
}


- (void)setupViews {
    _view_W = CGRectGetWidth(self.frame);
    _view_H = CGRectGetHeight(self.frame);
   
//    //menuView
//    self.titles = [self.dataSoure menuViewTitlesForTopScrollMenuView: self];
////    _menuView = [SGTopScrollMenu topScrollMenuWithFrame:CGRectMake(0, 0, _view_W - button_W -1 , menuView_H)];
//    if (self.titles.count > 4) {
//        _menuView = [SGTopScrollMenu topScrollMenuWithFrame:CGRectMake(0, 0, _view_W, menuView_H)];
//        _menuView.scrollTitleArr = self.titles;
//    } else {
//        _menuView = [SGTopScrollMenu topScrollMenuWithFrame:CGRectMake(50, 0, _view_W -100, menuView_H)];
//        _menuView.staticTitleArr = self.titles;
//    }
//    _menuView.topScrollMenuDelegate = self;
//    _menuView.backgroundColor = [UIColor clearColor];
//    [self addSubview:_menuView];
//    
////    UIView *line = [[UIView alloc] initWithFrame:CGRectMake(CGRectGetMaxX(_menuView.frame), 10, 1, 20)];
////    line.backgroundColor = [UIColor grayColor];
////    [self addSubview:line];
////    
////    UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
////    button.frame = CGRectMake(CGRectGetMaxX(line.frame), 0, button_W, menuView_H);
////    [button setTitle:@"V" forState:UIControlStateNormal];
////    [button addTarget:self action:@selector(buttonAction) forControlEvents:UIControlEventTouchUpInside];
////    [self addSubview:button];
////    
////    
////    
    //bgScrollView
    NSInteger numerViews = [self.dataSoure numberOfViewsInTopScrollMenuView:self];
    if (numerViews != self.titles.count) {
        NSLog(@"TopScrollMenuView error:  views count != titles count !");
        return;
    }
    
    _bgScrollView = [[UIScrollView alloc] init];
    //_bgScrollView.frame = CGRectMake(0, menuView_H + 1, _view_W, _view_H);
    _bgScrollView.contentSize = CGSizeMake(self.frame.size.width * numerViews, 0);
    _bgScrollView.backgroundColor = [UIColor clearColor];
    _bgScrollView.pagingEnabled = YES;                      // 开启分页
    _bgScrollView.bounces = NO;                             // 没有弹簧效果
    _bgScrollView.showsHorizontalScrollIndicator = NO;      // 隐藏水平滚动条
    _bgScrollView.delegate = self;                          // 设置代理
    _bgScrollView.translatesAutoresizingMaskIntoConstraints = false;
    
    
    
//    [self addConstraint:<#(nonnull NSLayoutConstraint *)#>];
//    
//    navBarView.isHidden = !isShowNavBar();
//    self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[navBarView]-0-|", options: NSLayoutFormatOptions(), metrics: nil, views: views));
//    self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-[navBarView]", options: NSLayoutFormatOptions(), metrics: nil, views: views));
//    self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-[navBarView(64)]", options: NSLayoutFormatOptions(), metrics: nil, views: views));
    
    for (NSInteger i = 0 ; i < numerViews; i++ ) {
        UIView *view = [self.dataSoure TopScrollMenuView:self viewAtIndex:i];
        view.frame = CGRectMake(i*_view_W, 0, _view_W, _view_H);
        [_bgScrollView addSubview:view];
        [self.viewData addObject:view];
    }
    
    [self addSubview:_bgScrollView];

}

#pragma mark - - - menuView Left Button
//- (void)buttonAction {
//    if ([self.delegate respondsToSelector:@selector(selectButtonForTopScrollMenuView:)]) {
//        [self.delegate selectButtonForTopScrollMenuView:self];
//    }
//}


#pragma mark - - - SGTopScrollMenu代理方法

- (void)SGTopScrollMenu:(SGTopScrollMenu *)topScrollMenu didSelectTitleAtIndex:(NSInteger)index {
    // 1 计算滚动的位置
    CGFloat offsetX = index * _view_W;
    _bgScrollView.contentOffset = CGPointMake(offsetX, 0);
    if ([self.delegate respondsToSelector:@selector(topScrollMenuView:didSelectView:index:)]) {
        [self.delegate topScrollMenuView:self didSelectView:self.viewData[index] index:index];
    }
}

#pragma mark - UIScrollViewDelegate
- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView{
    
    // 1.计算滚动到哪一页
    NSInteger index = scrollView.contentOffset.x / scrollView.frame.size.width;
   
    // 2.把对应的标题选中
    UILabel *selLabel = _menuView.allTitleLabel[index];
   
    // 3.让选中的标题居中
    if (self.titles.count > 4) {
        [_menuView scrollTitleLabelSelecteded:selLabel];
        [_menuView scrollTitleLabelSelectededCenter:selLabel];
    } else {
        [_menuView staticTitleLabelSelecteded:selLabel];

    }
    
    if ([self.delegate respondsToSelector:@selector(topScrollMenuView:didSelectView:index:)]) {
        [self.delegate topScrollMenuView:self didSelectView:self.viewData[index] index:index];
    }
}


@end

File: askdog-apps/iOS/AskDog/AskDog/Classes/App/UI/Home/HYTopScrollMenuView.swift
//
//  HYTopScrollMenuView.swift
//  AskDog
//
//  Created by Symond on 16/9/20.
//  Copyright © 2016年 Hooying. All rights reserved.
//

import UIKit

protocol HYTopScrollMenuViewDelegate  :class{
    func topScrollMenuView(menuView mv:HYTopScrollMenuView,didSelectedAtInndex ins:Int) -> Void;
    
}

protocol HYTopScrollMenuViewDataSource  :class{
    func numberOfViewsInTopScrollMenuView(menuView mv:HYTopScrollMenuView) -> Int ;
    func menuViewTitlesForTopScrollMenuView(menuView mv:HYTopScrollMenuView) -> [Any] ;
    func topScrollMenuView(menuView mv:HYTopScrollMenuView,index ins:Int) -> UIView;
    
}




class HYTopScrollMenuView: UIView,SGTopScrollMenuDelegate,UIScrollViewDelegate{
    
    weak var dataSource:HYTopScrollMenuViewDataSource?
    weak var delegate:HYTopScrollMenuViewDelegate?
    
    var bgScrollView:UIScrollView!
    var titles:[Any]!
    var menuView:SGTopScrollMenu!
    var views:[UIView] = [UIView]();
    
    
    override init(frame: CGRect) {
        super.init(frame: frame);
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder);
        //self.setupViews();
    }
    
    override func layoutSubviews() {
        super.layoutSubviews();
        
        print("\(self.frame.size.width)  \(self.frame.size.height)");
        //self.setupViews();
    }
    
    func setupViews() -> Void {
        
        self.titles = dataSource?.menuViewTitlesForTopScrollMenuView(menuView: self);
        
        menuView = SGTopScrollMenu(frame: CGRect(x: 0, y: 0, width: self.frame.size.width, height: 44));
        //self.menuView = SGTopScrollMenu(frame: CGRect.zero);
        //self.menuView.translatesAutoresizingMaskIntoConstraints = false;
        menuView.staticTitleArr = titles;
        menuView.backgroundColor = UIColor.clear;
        menuView.topScrollMenuDelegate = self;
        self.addSubview(menuView);
        
        let numberViews:Int = (dataSource?.numberOfViewsInTopScrollMenuView(menuView: self))!;
//        if(numberViews != self.titles.count){
//            print("TopScrollMenuView error:  views count != titles count !");
//            return;
//        }
        
        self.bgScrollView = UIScrollView(frame: CGRect.zero);
        self.bgScrollView.translatesAutoresizingMaskIntoConstraints = false;
        self.bgScrollView.backgroundColor = UIColor.clear;
        self.bgScrollView.isPagingEnabled = true;
        self.bgScrollView.bounces = false;
        self.bgScrollView.showsHorizontalScrollIndicator = false;
        self.bgScrollView.delegate = self;
        self.bgScrollView.contentSize = CGSize(width: self.frame.size.width * CGFloat(numberViews), height: 0);
        self.addSubview(self.bgScrollView);
        
        var views:[String:AnyObject] = [String:AnyObject]();
        //views["menuView"] = menuView;
        views["bgScrollView"] = bgScrollView;
        //self.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[menuView]-0-|", options: NSLayoutFormatOptions(), metrics: nil, views: views));
        //self.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-0-[menuView(44)]", options: NSLayoutFormatOptions(), metrics: nil, views: views));
        self.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[bgScrollView]-0-|", options: NSLayoutFormatOptions(), metrics: nil, views: views));
        self.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-44-[bgScrollView]-0-|", options: NSLayoutFormatOptions(), metrics: nil, views: views));
        
        for index in 0 ..< numberViews
        {
            let v:UIView = (dataSource?.topScrollMenuView(menuView: self, index: index))!;
            v.frame = CGRect(x: CGFloat(index) * self.frame.size.width, y: 0, width: self.frame.size.width, height: self.frame.size.height);
            self.bgScrollView.addSubview(v);
            self.views.append(v);
            
        }
    }

    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */
    
    //MARK:SGTopScrollMenuDelegate
    
    func sgTopScrollMenu(_ topScrollMenu: SGTopScrollMenu!, didSelectTitleAt index: Int) {
        let offsetX:CGFloat = CGFloat(index) * self.frame.size.width;
        self.bgScrollView.contentOffset = CGPoint(x: offsetX, y: 0);
        delegate?.topScrollMenuView(menuView: self, didSelectedAtInndex: index);
    }
    
    //MARK:UIScrollViewDelegate
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        let nIndex:Int = Int(scrollView.contentOffset.x) / Int(scrollView.frame.size.width);
        let lbl:UILabel = self.menuView.allTitleLabel[nIndex] as! UILabel;
        
        if(self.titles.count > 4){
            self.menuView.scrollTitleLabelSelecteded(lbl);
            self.menuView.scrollTitleLabelSelectededCenter(lbl);
        }else{
            self.menuView.staticTitleLabelSelecteded(lbl);
        }
        
        delegate?.topScrollMenuView(menuView: self, didSelectedAtInndex: nIndex);
        
    }
    
    func reloadTopScrollView() -> Void {
        self.menuView.removeFromSuperview();
        self.bgScrollView.removeFromSuperview();
       // self.views.removeAll();
       // self.setNeedsLayout();
    }
    
//    func reloadTopScrollView() -> Void {
//        for v in self.views{
//            v.removeFromSuperview();
//        }
//        
//        let numberViews:Int = (dataSource?.numberOfViewsInTopScrollMenuView(menuView: self))!;
//        self.bgScrollView.contentSize = CGSize(width: self.frame.size.width * CGFloat(numberViews), height: 0);
//        
//        for index in 0 ..< numberViews
//        {
//            let v:UIView = (dataSource?.topScrollMenuView(menuView: self, index: index))!;
//            v.frame = CGRect(x: CGFloat(index) * self.frame.size.width, y: 0, width: self.frame.size.width, height: self.frame.size.height);
//            self.bgScrollView.addSubview(v);
//            self.views.append(v);
//            
//        }
//    }
   
    

}
